### PR TITLE
fix ec2 subnet tags

### LINKED ISSUE #1113: ec2_vpc module will not update subnet tags
#### Issue Type:

Bug Report
##### Ansible Version:

ansible 1.9.0.1
##### Environment:

n/a
##### Summary:

When declaring a VPC with existing subnets that match the declared CIDR blocks, but have _different tags_, the module ignores the tags.

(Looking at the code, the same thing seems to be the case for AZs, but I haven't run into that issue myself.)
##### Steps To Reproduce:

Assuming we've got a VPC with the `10.0.0.0/16` block, tagged `Name="myvpc"`. The VPC has 1 subnet for the `10.0.1.0/24` block, tagged `Name="wrongname"`.

Running this task:

```
- ec2_vpc:
  cidr_block: 10.0.0.0/16
  region: "{{ ec2_region }}"
  resource_tags:
    Name: myvpc
  subnets:
    - cidr: 10.0.1.0/24
      az: "{{ ec2_availability_zone }}"
      resource_tags:
        Name: great_subnet_name
```

will not change the tag value for the subnet.  
##### Expected Results:

After running the task, the `10.0.1.0/24` subnet should have the tag `Name="great_subnet_name"`.
##### Actual Results:

It has the tag `Name="wrongname"`.

### PR DESCRIPTION
**Pull request by autotune**
This was created as a new branch in order to allow a pull request while pending review, so I can continue submitting a PR for a different feature as well. See comments  located at: https://github.com/ansible/ansible-modules-core/pull/2713.

fixes #1113

### CODE REVIEW

File: cloud/amazon/ec2_vpc.py [hunk 1]
```diff
@@ -406,9 +407,37 @@ def create_vpc(module, vpc_conn):
...
+            new_subnet_tags = subnet.get('resource_tags', None)
+            subnet_tags_delete = []
+            
             for csn in current_subnets:
                 if subnet['cidr'] == csn.cidr_block:
                     add_subnet = False
+
+                # Remove any subnet tags not in playbook
+                subnet_tags_extra = (set(csn.tags.items()).issubset(set(new_subnet_tags.items())))
+                if subnet_tags_extra == False:
```
Comment (carsongee) [inline cloud/amazon/ec2_vpc.py]: I would change this to `if not subnet_tags_extra`

File: cloud/amazon/ec2_vpc.py [hunk 2]
```diff
@@ -406,9 +407,37 @@ def create_vpc(module, vpc_conn):
...
+                                subnet_tags_delete.append(item)
+
+                        subnet_tags_delete = [key[0] for key in subnet_tags_delete]
+                        delete_subnet_tag = vpc_conn.delete_tags(csn.id, subnet_tags_delete)
+                        changed = True
+                    except EC2ResponseError, e:
+                        module.fail_json(msg='Unable to delete resource tag, error {0}'.format(e))
+                # Add new subnet tags if not current
+                subnet_tags_current = (set(new_subnet_tags.items()).issubset(set(csn.tags.items())))
+                if subnet_tags_current != True:
```
Comment (carsongee) [inline cloud/amazon/ec2_vpc.py]: I would maybe move the `subnet_tags_current` up to above the if and just do this block as an else statement

**Comment by carsongee** [comment]
Other than the two small changes, this looks good.

File: cloud/amazon/ec2_vpc.py (hunk 2 above)
Comment (autotune) [inline cloud/amazon/ec2_vpc.py]: Awesome. So just to clarify, the intended context is that the subnet tags are deleted only if there are no tags to be added?

Comment (carsongee) [inline cloud/amazon/ec2_vpc.py]: oh, I misread the logic there, sorry about that.  This is right. For pep8 though, it should be `if subnet_tags_current is not True:` or `if not tags_current` and `if not subnet_tags_extra` or `if subnet_tags_extra is False`.

Comment (autotune) [inline cloud/amazon/ec2_vpc.py]: No worries, went ahead and made that fix and thanks for the info about pep8. I'll be giving that a read.

**Comment by robynbergeron** [comment]
Thanks @autotune. @carsongee please review according to guidelines (http://docs.ansible.com/ansible/developing_modules.html#module-checklist) and comment with text 'shipit' or 'needs_revision' as appropriate.

File: cloud/amazon/ec2_vpc.py [hunk 3]
```diff
@@ -406,9 +407,39 @@ def create_vpc(module, vpc_conn):
...
+            
             for csn in current_subnets:
                 if subnet['cidr'] == csn.cidr_block:
                     add_subnet = False
+
+                # Check if AWS subnet tags are in playbook subnet tags
+                subnet_tags_extra = (set(csn.tags.items()).issubset(set(new_subnet_tags.items())))
+                # Check if subnet tags in playbook are in AWS subnet tags 
+                subnet_tags_current = (set(new_subnet_tags.items()).issubset(set(csn.tags.items())))
+                if subnet_tags_extra == False:
```
Comment (carsongee) [inline cloud/amazon/ec2_vpc.py]: one last tiny nitpick: this should be `if subnet_tags_extra is False` per pep8

**Comment by carsongee** [comment]
shipit and thanks so much @autotune

**Comment by gregdek** [comment]
Thanks again to @autotune for this PR, and thanks @carsongee for reviewing. Marking for inclusion.

**Comment by bcoca** [comment]
@autotune  could you flatten the commit history?

**Comment by autotune** [comment]
@bcoca done!

**Pull request merged**